File: backend/app/api/routes/goals.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from ... import schemas
from ...services.goals import (
    create_goal as create_goal_svc,
    delete_goal,
    get_goal,
    get_goals_with_progress,
    list_goals,
    update_goal as update_goal_svc,
)
from ..deps import get_current_user, get_db_session

router = APIRouter(prefix="/goals", tags=["goals"])


@router.get("", response_model=schemas.GoalsProgressResponse)
def get_goals(
    period: str = Query("7d", description="Progress period: 7d, month, deadline"),
    include_archived: bool = Query(False, description="Include archived goals"),
    db: Session = Depends(get_db_session),
    user=Depends(get_current_user),
):
    if period not in schemas.GOAL_PROGRESS_PERIODS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"period must be one of {schemas.GOAL_PROGRESS_PERIODS}",
        )
    return get_goals_with_progress(db, user.id, period=period, include_archived=include_archived)


@router.post("", response_model=schemas.GoalRead, status_code=status.HTTP_201_CREATED)
def create_goal(
    payload: schemas.GoalCreate,
    db: Session = Depends(get_db_session),
    user=Depends(get_current_user),
):
    if payload.sphere not in schemas.GOAL_SPHERES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"sphere must be one of {schemas.GOAL_SPHERES}",
        )
    goals = list_goals(db, user.id, include_archived=False)
    if len(goals) >= schemas.GOAL_MAX_ACTIVE:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Maximum {schemas.GOAL_MAX_ACTIVE} active goals. Archive or delete one first.",
        )
    return create_goal_svc(db, user.id, payload)


@router.get("/{goal_id}", response_model=schemas.GoalRead)
def read_goal(
    goal_id: int,
    db: Session = Depends(get_db_session),
    user=Depends(get_current_user),
):
    goal = get_goal(db, goal_id, user.id)
    if not goal:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Goal not found")
    return goal


@router.patch("/{goal_id}", response_model=schemas.GoalRead)
def update_goal(
    goal_id: int,
    payload: schemas.GoalUpdate,
    db: Session = Depends(get_db_session),
    user=Depends(get_current_user),
):
    goal = get_goal(db, goal_id, user.id)
    if not goal:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Goal not found")
    if payload.sphere is not None and payload.sphere not in schemas.GOAL_SPHERES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"sphere must be one of {schemas.GOAL_SPHERES}",
        )
    return update_goal_svc(db, goal, payload)


@router.delete("/{goal_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_goal(
    goal_id: int,
    db: Session = Depends(get_db_session),
    user=Depends(get_current_user),
):
    goal = get_goal(db, goal_id, user.id)
    if not goal:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Goal not found")
    delete_goal(db, goal)

File: frontend-react/src/shared/api/goals.ts
import api from './client'

export const GOAL_SPHERES = ['health', 'finance', 'productivity', 'learning'] as const
export type GoalSphere = (typeof GOAL_SPHERES)[number]

export type Goal = {
  id: number
  user_id: number
  sphere: string
  title: string
  target_value?: number | null
  target_metric?: string | null
  deadline?: string | null
  created_at: string
}

export type GoalProgress = {
  goal_id: number
  title: string
  sphere: string
  target_value?: number | null
  target_metric?: string | null
  current_value?: number | null
  progress_pct?: number | null
  deadline?: string | null
}

export type GoalsProgressResponse = {
  goals: Goal[]
  progress: GoalProgress[]
}

export const getGoals = () =>
  api.get<GoalsProgressResponse>('/goals').then((res) => res.data)

export const createGoal = (payload: {
  sphere: string
  title: string
  target_value?: number | null
  target_metric?: string | null
  deadline?: string | null
}) => api.post<Goal>('/goals', payload).then((res) => res.data)

export const updateGoal = (
  goalId: number,
  payload: Partial<{
    sphere: string
    title: string
    target_value: number | null
    target_metric: string | null
    deadline: string | null
  }>,
) => api.patch<Goal>(`/goals/${goalId}`, payload).then((res) => res.data)

export const deleteGoal = (goalId: number) =>
  api.delete(`/goals/${goalId}`)
